pkg/arvo/lib/shoe.hoon

```diff
@@ -0,0 +1,341 @@
+::  shoe: console application library
+::
+::    /lib/sole: draw some characters
+::    /lib/shoe: draw the rest of the fscking app
+::
+::    call +agent with a type, then call the resulting function with a core
+::    of the shape described in +shoe.
+::    you may produce classic gall cards and "shoe-effects", shorthands for
+::    sending cli events to connected clients.
+::    default implementations for the shoe-specific arms are in +default.
+::    for a simple usage example, see /app/shoe.
+::
+/-  *sole
+/+  sole, auto=language-server-complete
+|%
++$  state-0
+  $:  %0
+      soles=(map @ta sole-share)
+  ==
+::  $card: standard gall cards plus shoe effects
+::
++$  card
+  $%  card:agent:gall
+      [%shoe sole-ids=(list @ta) effect=shoe-effect]  ::  ~ sends to all soles
+  ==
+::  $shoe-effect: easier sole-effects
+::
++$  shoe-effect
+  $%  [%sole effect=sole-effect]
+      ::TODO  complex screen-draw effects
+  ==
+::  +shoe: gall agent core with extra arms
+::
+++  shoe
+  |*  command-type=mold
+  $_  ^|
+  |_  bowl:gall
+  ++  command-parser
+    |~  sole-id=@ta
+    |~(nail *(like command-type))
+  ::
+  ++  tab-list
+    |~  sole-id=@ta
+    ::  (list [@t tank])
+    *(list (option:auto tank))
+  ::
+  ++  on-command
+    |~  [sole-id=@ta command=command-type]
+    *(quip card _^|(..on-init))
+  ::
+  ++  can-connect
```

Comment:
I like requiring them to implement `+can-connect` like this.  Makes you have to *specifically* say "I'm open to communication from other ships".  I don't like that our other arms fail open, maybe we should do something similar for those eventually.